File: tests/node-bindgen/ts/error_handeling.ts
import { Group } from "./common";
import { StructErrorHandeling } from "binding";

const tests = new Group("Error Handeling Tests");
const struct = new StructErrorHandeling();

{
    const test = tests.test("testOfException");
    try {
        const _ = struct.testOfException();
        test.fail("function should throw exception");
    } catch (err) {
        test.assert(err).msg("Value of err invalid").type("string");
        test.success();
    }
}

{
    const test = tests.test("testOfExceptionSuppression");
    const err = struct.testOfExceptionSuppression();
    if (err instanceof Error) {
        test.assert(err.message).msg("Value of err invalid").equal("test");
    } else {
        test.fail("Value of err invalid");
    }
    test.success();
}

{
    const test = tests.test("testOfExceptionWithCustomError");
    try {
        const _ = struct.testOfExceptionWithCustomError();
        test.fail("function should throw exception");
    } catch (err) {
        if (err instanceof Error) {
            test.assert((err as any).err)
                .msg("Value of err invalid")
                .type("object");
        } else {
            test.fail("Value of err invalid");
        }
        test.success();
    }
}

{
    const test = tests.test("testOfExceptionSuppressionWithCustomError");
    try {
        const err = struct.testOfExceptionSuppressionWithCustomError();
        if (err instanceof Error) {
            test.assert(err.err).msg("Value of err invalid").type("object");
        } else {
            test.fail("Value of err invalid");
        }
        test.success();
    } catch (err) {
        test.fail("function should not throw exception");
    }
}

File: tests/node-bindgen/rs/error_handeling.rs
use node_bindgen::derive::node_bindgen;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tslink::tslink;

#[tslink]
#[derive(Serialize, Deserialize)]
struct ErrorC {
    msg: String,
    code: usize,
}

impl From<serde_json::Error> for ErrorC {
    fn from(value: serde_json::Error) -> Self {
        ErrorC {
            msg: value.to_string(),
            code: 1,
        }
    }
}

struct StructErrorHandeling {}

#[tslink(class)]
#[node_bindgen]
impl StructErrorHandeling {
    #[tslink(constructor)]
    #[node_bindgen(constructor)]
    pub fn new() -> Self {
        Self {}
    }

    #[tslink(snake_case_naming, exception_suppression)]
    #[node_bindgen]
    fn test_of_exception_suppression(&self) -> Result<i32, String> {
        Err("test".to_string())
    }

    #[tslink(error = "json", snake_case_naming, exception_suppression)]
    #[node_bindgen]
    fn test_of_exception_suppression_with_custom_error(&self) -> Result<i32, ErrorC> {
        Err(ErrorC {
            msg: "test".to_string(),
            code: 666,
        })
    }

    #[tslink(snake_case_naming)]
    #[node_bindgen]
    fn test_of_exception(&self) -> Result<i32, String> {
        Err("test".to_string())
    }

    #[tslink(error = "json", snake_case_naming)]
    #[node_bindgen]
    fn test_of_exception_with_custom_error(&self) -> Result<i32, ErrorC> {
        Err(ErrorC {
            msg: "test".to_string(),
            code: 666,
        })
    }
}
